src/webgpu/shader/execution/robust_access.spec.ts

```diff
@@ -0,0 +1,426 @@
+    const nextErrorReturnValue = (() => {
+      let errorReturnValue = 0x1000;
+      return () => {
+        ++errorReturnValue;
+        return `0x${errorReturnValue.toString(16)}u`;
+      };
+    })();
+
+    // Produce the accesses to the variable.
+    for (const indexToTest of indicesToTest) {
+      // TODO check with constants too if WGSL allows it.
+      // FIXME: ^ What does this comment mean, does it mean pipeline-overridable constants?
+      //          Move this TODO into the test description once that's figured out.
+      const index = `(${indexToTest}) + ${constZeroIndex}`;
+      const zeroElement = `${_typeInfo.elementType}()`;
+
+      switch (access) {
+        case 'read':
+          {
+            let element = `s.data[${index}]`;
+            if (atomic) element = `atomicLoad(&${element})`;
+            let condition = `${element} != ${zeroElement}`;
+            if ('innerLength' in _typeInfo) condition = `any(${condition})`;
+            testFunctionSource += `
+            if (${condition}) { return ${nextErrorReturnValue()}; }`;
+          }
+          break;
+
+        case 'write':
+          if (atomic) {
+            testFunctionSource += `
+              atomicStore(&s.data[${index}], ${zeroElement});`;
+          } else {
+            testFunctionSource += `
+              s.data[${index}] = ${zeroElement};`;
+          }
+          break;
+      }
+    }
+
+    // Check that the canaries haven't been modified
+    if (usesCanary) {
+      testFunctionSource += `
+        for (var i = 0u; i < 10u; i = i + 1u) {
+          if (s.startCanary[i] != 0xFFFFFFFFu) {
+            return ${nextErrorReturnValue()};
+          }
+          if (s.endCanary[i] != 0xFFFFFFFFu) {
+            return ${nextErrorReturnValue()};
+          }
+        }`;
+    }
+
+    // Run the test
+
+    // First aggregate the test source
+    const testSource = `
+      ${globalSource}
+
+      fn runTest() -> u32 {
+        ${testFunctionSource}
+        return 0u;
+      }`;
+
+    // Run it.
+    if (bufferBindingSize !== undefined) {
+      assert(baseType !== 'bool', 'case should have been filtered out');
+
+      const expectedData = new ArrayBuffer(testBufferSize);
+      const bufferBindingEnd = bufferBindingOffset + bufferBindingSize;
+      testFillArrayBuffer(expectedData, baseType, {
+        zeroByteStart: bufferBindingOffset,
+        zeroByteCount: bufferBindingSize,
+      });
+
+      // Create a buffer that contains zeroes in the allowed access area, and 42s everywhere else.
+      const testBuffer = t.device.createBuffer({
+        mappedAtCreation: true,
+        size: testBufferSize,
+        usage:
+          GPUBufferUsage.COPY_SRC |
+          GPUBufferUsage.UNIFORM |
+          GPUBufferUsage.STORAGE |
+          GPUBufferUsage.COPY_DST,
+      });
+      memcpy({ src: expectedData, dst: testBuffer.getMappedRange() });
```

Comment:
Maybe we can call testFillArrayBuffer directly on getMappedRange?

Reply:
I realized I could just use `t.makeBufferWithContents(expectedData, usage)`